Source/RequestTaskMap.swift

```diff
@@ -83,10 +91,46 @@ struct RequestTaskMap {
         return tasksToRequests.count
     }
 
+    var eventCount: Int {
+        precondition(taskEvents.count == count, "RequestTaskMap.eventCount invalid, count: \(count) != taskEvents.count: \(taskEvents.count)")
+
+        return taskEvents.count
+    }
+
     var isEmpty: Bool {
         precondition(tasksToRequests.isEmpty == requestsToTasks.isEmpty,
                      "RequestTaskMap.isEmpty invalid, requests.isEmpty: \(tasksToRequests.isEmpty) != tasks.isEmpty: \(requestsToTasks.isEmpty)")
 
         return tasksToRequests.isEmpty
     }
+
+    var isEventsEmpty: Bool {
+        precondition(taskEvents.isEmpty == isEmpty, "RequestTaskMap.isEventsEmpty invalid, isEmpty: \(isEmpty) != taskEvents.isEmpty: \(taskEvents.isEmpty)")
+
+        return taskEvents.isEmpty
+    }
+
+    mutating func disassociateIfNecessaryAfterGatheringMetricsForTask(_ task: URLSessionTask) {
```

Comment:
I wonder if reusing the state provider APIs would make sense here:

```swift
disassociateIfNecessaryAfterGatheringMetricsForTask -> didGatherMetricsForTask
disassociateIfNecessaryAfterCompletingTask -> didCompleteTask
```

Reply:
Those were actually my first names for these methods but I wanted to indicate there was a side effect of possibly breaking the association between task and request.

Reply:
Gotcha. I think I'd roll with these then.